# Topic: **Sorting, Ranking, Indexes with Duplicate Labels, `loc` and `iloc`**


In [384]:
import pandas as pd
import numpy as np

In [385]:
data = pd.read_csv('material.csv')

In [386]:
df = pd.DataFrame(data)

In [387]:
df.head()

Material   Su   Sy       E      G   mu    Ro   Use
0   ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.3  7860  True
1  ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.3  7860  True
2    ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.3  7860  True
3   ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.3  7860  True
4  ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.3  7860  True

### Sorting

1. Sort the dataset by `Su` in descending order and display the top 10 materials.

In [388]:
df.sort_values(by=['Su'], ascending=False).head(10)

Material    Su    Sy       E       G   mu  \
86     ANSI Steel SAE 5160 tempered at 400 F  2220  1793  207000   79000  0.3   
105    ANSI Steel SAE 9255 tempered at 400 F  2103  2048  207000   79000  0.3   
102    ANSI Steel SAE 8740 tempered at 400 F  1999  1655  207000   79000  0.3   
156  ANSI Steel SAE 51440C tempered at 600 F  1965  1896  200000  769000  0.3   
83     ANSI Steel SAE 5150 tempered at 400 F  1944  1731  207000   79000  0.3   
98     ANSI Steel SAE 8650 tempered at 400 F  1937  1675  207000   79000  0.3   
66     ANSI Steel SAE 4150 tempered at 400 F  1931  1724  207000   79000  0.3   
89     ANSI Steel SAE 6150 tempered at 400 F  1931  1689  207000   79000  0.3   
71     ANSI Steel SAE 4340 tempered at 400 F  1875  1675  207000   79000  0.3   
95     ANSI Steel SAE 8640 tempered at 400 F  1862  1669  207000   79000  0.3   

       Ro    Use  
86   7860  False  
105  7860  False  
102  7860  False  
156  8030  False  
83   7860  False  
98   7860  False  
66   7860  False  
89   7860  False  
71   7860  False  
95   7860  False

2. Sort the dataset by `Use` and then by `Sy` in ascending order.

In [389]:
df = pd.DataFrame(data)

In [390]:
df.sort_values(by=['Use', 'Sy'], ascending=True).head(10)

Material   Su  Sy       E      G    mu  \
382         ANSI Aluminum Alloy 1060-O Wrought   69  28   73000  26000  0.33   
392         ANSI Aluminum Alloy 1350-O Wrought   83  28   73000  26000  0.33   
387         ANSI Aluminum Alloy 1100-O Wrought   90  34   73000  26000  0.33   
254                  ANSI Copper Alloy C15000   200  41  108000  40000  0.35   
364        ANSI Aluminum Alloy B443.0 Cast (F)  117  41   73000  26000  0.33   
440         ANSI Aluminum Alloy 3003-O Wrought  110  41   73000  26000  0.33   
445  ANSI Aluminum Alloy Alclad 3003-O Wrought  110  41   73000  26000  0.33   
467         ANSI Aluminum Alloy 5005-O Wrought  124  41   73000  26000  0.33   
259                  ANSI Copper Alloy C16200   241  48  108000  40000  0.35   
363         ANSI Aluminum Alloy 443.0 Cast (F)  117  48   73000  26000  0.33   

       Ro    Use  
382  2700  False  
392  2700  False  
387  2700  False  
254  8930  False  
364  2700  False  
440  2700  False  
445  2700  False  
467  2700  False  
259  8930  False  
363  2700  False

3. Sort the dataset by the alphabetical order of the `Material` column and show the last 5 rows.

In [391]:
df = pd.DataFrame(data)

In [392]:
df.sort_values(by=['Material'], ascending=True).tail(5)

Material   Su   Sy       E      G   mu    Ro    Use
1398             NF NF ZES   270  225  206000  80000  0.3  7860  False
1481  NF Nodular cast iron   700  420  169000  70000  0.2  7160  False
1482  NF Nodular cast iron   800  480  169000  70000  0.2  7160  False
1480  NF Nodular cast iron   600  370  169000  70000  0.2  7160  False
1479  NF Nodular cast iron   500  320  169000  70000  0.2  7160  False

4. Sort the DataFrame by the index after setting `Material` as the index.

In [393]:
df = pd.DataFrame(data)

In [394]:
df.set_index('Material', inplace=True)

In [395]:
df.sort_index()

Su   Sy       E      G    mu    Ro  \
Material                                                                    
ANSI Aluminum Alloy 1060-H12 Wrought   83   76   73000  26000  0.33  2700   
ANSI Aluminum Alloy 1060-H14 Wrought   97   90   73000  26000  0.33  2700   
ANSI Aluminum Alloy 1060-H16 Wrought  110  103   73000  26000  0.33  2700   
ANSI Aluminum Alloy 1060-H18 Wrought  131  124   73000  26000  0.33  2700   
ANSI Aluminum Alloy 1060-O Wrought     69   28   73000  26000  0.33  2700   
...                                   ...  ...     ...    ...   ...   ...   
NF NF ZES                             270  225  206000  80000  0.30  7860   
NF Nodular cast iron                  700  420  169000  70000  0.20  7160   
NF Nodular cast iron                  800  480  169000  70000  0.20  7160   
NF Nodular cast iron                  600  370  169000  70000  0.20  7160   
NF Nodular cast iron                  500  320  169000  70000  0.20  7160   

                                        Use  
Material                                     
ANSI Aluminum Alloy 1060-H12 Wrought  False  
ANSI Aluminum Alloy 1060-H14 Wrought  False  
ANSI Aluminum Alloy 1060-H16 Wrought  False  
ANSI Aluminum Alloy 1060-H18 Wrought  False  
ANSI Aluminum Alloy 1060-O Wrought    False  
...                                     ...  
NF NF ZES                             False  
NF Nodular cast iron                  False  
NF Nodular cast iron                  False  
NF Nodular cast iron                  False  
NF Nodular cast iron                  False  

[1552 rows x 7 columns]

5. Sort the rows by column `Ro`, placing `NaN` (if any) at the top.

In [396]:
df = pd.DataFrame(data)

In [397]:
df.sort_values(by=['Ro'], na_position='first')

Material   Su   Sy       E      G  \
585  ANSI Magnesium Alloy AM100A-T61 sand casting  276  152   73000  26000   
586    ANSI Magnesium Alloy AZ63A-T6 sand casting  276   97   73000  26000   
587    ANSI Magnesium Alloy AZ81A-T4 sand casting  276   83   73000  26000   
588    ANSI Magnesium Alloy AZ91C-T6 sand casting  276  145   73000  26000   
589    ANSI Magnesium Alloy AZ91E-T6 sand casting  276  145   73000  26000   
..                                            ...  ...  ...     ...    ...   
256                     ANSI Copper Alloy C15715   359  303  108000  40000   
257                     ANSI Copper Alloy C15720   441  372  108000  40000   
253                     ANSI Copper Alloy C10100   221   69  108000  40000   
270                     ANSI Copper Alloy C19500   552  448  108000  40000   
271                     ANSI Copper Alloy C21000   234   69  108000  40000   

       mu    Ro    Use  
585  0.33  1750  False  
586  0.33  1750  False  
587  0.33  1750  False  
588  0.33  1750  False  
589  0.33  1750  False  
..    ...   ...    ...  
256  0.35  8930  False  
257  0.35  8930  False  
253  0.35  8930  False  
270  0.35  8930  False  
271  0.35  8930  False  

[1552 rows x 8 columns]

### Ranking

6. Add a column `'Strength Rank'` that ranks materials by `Su` in descending order.

In [398]:
df = pd.DataFrame(data)

In [399]:
df['Strength_Rank'] = df['Su'].rank(ascending=False)

In [400]:
df[['Material', 'Su', 'Strength_Rank']]

Material   Su  Strength_Rank
0      ANSI Steel SAE 1015 as-rolled  421         1002.5
1     ANSI Steel SAE 1015 normalized  424          999.0
2       ANSI Steel SAE 1015 annealed  386         1084.5
3      ANSI Steel SAE 1020 as-rolled  448          948.0
4     ANSI Steel SAE 1020 normalized  441          957.0
...                              ...  ...            ...
1547          JIS Nodular cast iron   600          573.0
1548          JIS Nodular cast iron   700          404.0
1549          JIS Nodular cast iron   800          294.0
1550        JIS Malleable cast iron   400         1058.5
1551        JIS Malleable cast iron   500          793.0

[1552 rows x 3 columns]

In [401]:
# integer rank
df['Strength_Rank'] = df['Su'].rank(ascending=False, method='min').astype(int)

In [402]:
df[['Material', 'Su', 'Strength_Rank']]

Material   Su  Strength_Rank
0      ANSI Steel SAE 1015 as-rolled  421           1001
1     ANSI Steel SAE 1015 normalized  424            999
2       ANSI Steel SAE 1015 annealed  386           1083
3      ANSI Steel SAE 1020 as-rolled  448            945
4     ANSI Steel SAE 1020 normalized  441            952
...                              ...  ...            ...
1547          JIS Nodular cast iron   600            561
1548          JIS Nodular cast iron   700            394
1549          JIS Nodular cast iron   800            281
1550        JIS Malleable cast iron   400           1047
1551        JIS Malleable cast iron   500            776

[1552 rows x 3 columns]

7. Rank the materials based on `E`, using the `'dense'` method.

In [403]:
df = pd.DataFrame(data)

In [404]:
df['e_rank'] = df['E'].rank(method='dense').astype(int)

In [405]:
df[['Material', 'E', 'e_rank']]

Material       E  e_rank
0      ANSI Steel SAE 1015 as-rolled  207000      34
1     ANSI Steel SAE 1015 normalized  207000      34
2       ANSI Steel SAE 1015 annealed  207000      34
3      ANSI Steel SAE 1020 as-rolled  207000      34
4     ANSI Steel SAE 1020 normalized  207000      34
...                              ...     ...     ...
1547          JIS Nodular cast iron   169000      20
1548          JIS Nodular cast iron   169000      20
1549          JIS Nodular cast iron   169000      20
1550        JIS Malleable cast iron   160000      19
1551        JIS Malleable cast iron   160000      19

[1552 rows x 3 columns]

8. Rank materials by `G`, breaking ties using `'first'` method.

In [406]:
df = pd.DataFrame(data)

In [407]:
df['g_rank'] = df['G'].rank(method='first').astype(int)

In [408]:
df[['Material', 'G', 'g_rank']]

Material      G  g_rank
0      ANSI Steel SAE 1015 as-rolled  79000     679
1     ANSI Steel SAE 1015 normalized  79000     680
2       ANSI Steel SAE 1015 annealed  79000     681
3      ANSI Steel SAE 1020 as-rolled  79000     682
4     ANSI Steel SAE 1020 normalized  79000     683
...                              ...    ...     ...
1547          JIS Nodular cast iron   70000     578
1548          JIS Nodular cast iron   70000     579
1549          JIS Nodular cast iron   70000     580
1550        JIS Malleable cast iron   64000     541
1551        JIS Malleable cast iron   64000     542

[1552 rows x 3 columns]

9. Create a column `'Combined Rank'` as the average of ranks for `Su`, `Sy`, and `E`.

In [409]:
df = pd.DataFrame(data)

In [410]:
df['su_rank'] = df['Su'].rank(method='first').astype(int)

In [411]:
df['sy_rank'] = df['Sy'].rank(method='first').astype(int)

In [412]:
df['e_rank'] = df['E'].rank(method='first').astype(int)

In [413]:
df['combined_rank'] = df[['su_rank', 'sy_rank', 'e_rank']].mean(axis=1).round(2)

In [414]:
df[['Material', 'Su', 'Sy', 'E', 'combined_rank']]

Material   Su   Sy       E  combined_rank
0      ANSI Steel SAE 1015 as-rolled  421  314  207000         897.67
1     ANSI Steel SAE 1015 normalized  424  324  207000         911.33
2       ANSI Steel SAE 1015 annealed  386  284  207000         840.67
3      ANSI Steel SAE 1020 as-rolled  448  331  207000         932.67
4     ANSI Steel SAE 1020 normalized  441  346  207000         949.33
...                              ...  ...  ...     ...            ...
1547          JIS Nodular cast iron   600  370  169000         851.33
1548          JIS Nodular cast iron   700  420  169000         942.00
1549          JIS Nodular cast iron   800  480  169000        1006.67
1550        JIS Malleable cast iron   400  180  160000         452.00
1551        JIS Malleable cast iron   500  260  160000         647.00

[1552 rows x 5 columns]

10. Get the top 5 materials based on `'Combined Rank'`.

In [415]:
df.sort_values(by=['combined_rank'], ascending=False).head(5)

Material    Su    Sy       E  \
1200  GOST Steel 40ChFA GOST 4543-71 Quenching and c...  1600  1300  215000   
105               ANSI Steel SAE 9255 tempered at 400 F  2103  2048  207000   
1218  GOST Steel 30ChGSA GOST 4543-71 Quenching and ...  1500  1300  215000   
1189  GOST Steel 35ChM GOST 4543-71 Quenching and co...  1600  1400  213000   
1149  GOST Steel 65G GOST 1050-88 Quenching and cool...  1500  1250  215000   

          G   mu    Ro    Use  su_rank  sy_rank  e_rank  combined_rank  
1200  83000  0.3  7810  False     1525     1519    1520        1521.33  
105   79000  0.3  7860  False     1551     1552    1459        1520.67  
1218  83000  0.3  7850  False     1519     1520    1523        1520.67  
1189  82000  0.3  7820  False     1524     1527    1498        1516.33  
1149  83000  0.3  7850  False     1517     1512    1518        1515.67

### Axis Indexes and Duplicates

11. Duplicate the first 5 rows and append them to the DataFrame. What is the output of `df.index.is_unique`?

In [416]:
df = pd.DataFrame(data)

In [417]:
df.index.is_unique

True

In [418]:
df_with_duplicates = pd.concat([df, df.iloc[:5]], ignore_index=False)

In [419]:
df_with_duplicates.head(5)

Material   Su   Sy       E      G   mu    Ro   Use
0   ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.3  7860  True
1  ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.3  7860  True
2    ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.3  7860  True
3   ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.3  7860  True
4  ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.3  7860  True

In [420]:
df_with_duplicates.tail(5)

Material   Su   Sy       E      G   mu    Ro   Use
0   ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.3  7860  True
1  ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.3  7860  True
2    ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.3  7860  True
3   ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.3  7860  True
4  ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.3  7860  True

In [421]:
df_with_duplicates.index.is_unique

False

12. After duplication, retrieve all rows with the duplicated index value (e.g., index 0).

In [422]:
df_with_duplicates.loc[[0, 1, 2, 3, 4]]

Material   Su   Sy       E      G   mu    Ro   Use
0   ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.3  7860  True
0   ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.3  7860  True
1  ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.3  7860  True
1  ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.3  7860  True
2    ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.3  7860  True
2    ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.3  7860  True
3   ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.3  7860  True
3   ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.3  7860  True
4  ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.3  7860  True
4  ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.3  7860  True

13. Reset the index and sort the DataFrame by `Material` again.

In [423]:
df_with_duplicates.reset_index()

index                        Material   Su   Sy       E      G   mu  \
0         0   ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.3   
1         1  ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.3   
2         2    ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.3   
3         3   ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.3   
4         4  ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.3   
...     ...                             ...  ...  ...     ...    ...  ...   
1552      0   ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.3   
1553      1  ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.3   
1554      2    ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.3   
1555      3   ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.3   
1556      4  ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.3   

        Ro   Use  
0     7860  True  
1     7860  True  
2     7860  True  
3     7860  True  
4     7860  True  
...    ...   ...  
1552  7860  True  
1553  7860  True  
1554  7860  True  
1555  7860  True  
1556  7860  True  

[1557 rows x 9 columns]

In [424]:
df_with_duplicates.sort_values(by=['Material'], ascending=True).head(5)

Material   Su   Sy      E      G    mu    Ro  \
383  ANSI Aluminum Alloy 1060-H12 Wrought   83   76  73000  26000  0.33  2700   
384  ANSI Aluminum Alloy 1060-H14 Wrought   97   90  73000  26000  0.33  2700   
385  ANSI Aluminum Alloy 1060-H16 Wrought  110  103  73000  26000  0.33  2700   
386  ANSI Aluminum Alloy 1060-H18 Wrought  131  124  73000  26000  0.33  2700   
382    ANSI Aluminum Alloy 1060-O Wrought   69   28  73000  26000  0.33  2700   

       Use  
383  False  
384  False  
385  False  
386  False  
382  False

14. Group the materials by `Use` and count how many rows are duplicates (check duplicated index).

In [425]:
df_with_duplicates['is_duplicate_index'] = df_with_duplicates.index.duplicated(keep=False)

In [426]:
duplicate_counts_by_use = df_with_duplicates[df_with_duplicates['is_duplicate_index']].groupby('Use').size()

In [427]:
duplicate_counts_by_use

Use
True    10
dtype: int64

### Explanation:
You duplicated the first 5 rows from the original DataFrame.

All those rows had 'Use' == True and shared index values.

Therefore, 10 rows total (5 original + 5 duplicates) have duplicated index values.

When grouped by the Use column:

All duplicates fall under Use = True.

15. Remove all rows with duplicate `Material` names while keeping the first occurrence.

In [428]:
df_unique_materials = df_with_duplicates.drop_duplicates(subset=['Material'], keep='first')

In [429]:
df_unique_materials

Material   Su   Sy       E      G    mu    Ro  \
0      ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.30  7860   
1     ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.30  7860   
2       ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.30  7860   
3      ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.30  7860   
4     ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.30  7860   
...                              ...  ...  ...     ...    ...   ...   ...   
1539                 JIS JIS SUH310   590  205  206000  80000  0.30  7860   
1540                  JIS JIS SUH35   880  560  206000  80000  0.30  7860   
1541             JIS Grey cast iron   150  150   85000  34000  0.25  7160   
1546          JIS Nodular cast iron   500  320  169000  70000  0.20  7160   
1550        JIS Malleable cast iron   400  180  160000  64000  0.27  7160   

        Use  is_duplicate_index  
0      True                True  
1      True                True  
2      True                True  
3      True                True  
4      True                True  
...     ...                 ...  
1539  False               False  
1540  False               False  
1541  False               False  
1546  False               False  
1550  False               False  

[1460 rows x 9 columns]

In [430]:
df_unique_materials.shape

(1460, 9)

This means 92 duplicates (from 1552 total) were removed based on the Material column.

In [431]:
df_unique_materials.index.is_unique

True

### `loc` and `iloc` Practice

16. Use `loc` to retrieve all materials with `Use == 'False'` and `Sy > 500`.

In [432]:
df = pd.DataFrame(data)

In [433]:
mask = (df['Use'] == False) & (df['Sy'] > 500)

In [434]:
df.loc[mask].head(5)

Material    Su   Sy       E      G   mu  \
12  ANSI Steel SAE 1030 tempered at 400 F   848  648  207000  79000  0.3   
16  ANSI Steel SAE 1040 tempered at 400 F   779  593  207000  79000  0.3   
20  ANSI Steel SAE 1050 tempered at 600 F   979  724  207000  79000  0.3   
24  ANSI Steel SAE 1060 tempered at 600 F  1103  779  207000  79000  0.3   
25          ANSI Steel SAE 1080 as-rolled   965  586  207000  79000  0.3   

      Ro    Use  
12  7860  False  
16  7860  False  
20  7860  False  
24  7860  False  
25  7860  False

17. Use `iloc` to extract the 10th to 20th rows and display only `Material`, `Su`, and `Use`.

In [435]:
df[['Material', 'Su', 'Use']].iloc[10:21]

Material   Su    Use
10         ANSI Steel SAE 1030 normalized  517  False
11           ANSI Steel SAE 1030 annealed  464   True
12  ANSI Steel SAE 1030 tempered at 400 F  848  False
13          ANSI Steel SAE 1040 as-rolled  621  False
14         ANSI Steel SAE 1040 normalized  590  False
15           ANSI Steel SAE 1040 annealed  519  False
16  ANSI Steel SAE 1040 tempered at 400 F  779  False
17          ANSI Steel SAE 1050 as-rolled  724  False
18         ANSI Steel SAE 1050 normalized  748  False
19           ANSI Steel SAE 1050 annealed  636  False
20  ANSI Steel SAE 1050 tempered at 600 F  979  False

18. Use `loc` to set all `Use == 'True'` rows where `Su < 400` to `False`.

In [436]:
df.loc[(df['Use'] == True) & (df['Su'] < 400), 'Use'] = False

In [437]:
df[['Su', 'Use']].head(5)

Su    Use
0  421   True
1  424   True
2  386  False
3  448   True
4  441   True

19. Use `iloc` to update the first 3 rows of `mu` to 0.33.

In [438]:
df = pd.DataFrame(data)

In [439]:
df.iloc[:3, df.columns.get_loc('mu')] = .33

In [440]:
df.head(5)

Material   Su   Sy       E      G    mu    Ro    Use
0   ANSI Steel SAE 1015 as-rolled  421  314  207000  79000  0.33  7860   True
1  ANSI Steel SAE 1015 normalized  424  324  207000  79000  0.33  7860   True
2    ANSI Steel SAE 1015 annealed  386  284  207000  79000  0.33  7860  False
3   ANSI Steel SAE 1020 as-rolled  448  331  207000  79000  0.30  7860   True
4  ANSI Steel SAE 1020 normalized  441  346  207000  79000  0.30  7860   True

20. Select and display the value of `E` for the row where `Material == 'ANSI Steel SAE 1015 as-rolled'` using `loc`.

In [441]:
df.loc[(df['Material'] == 'ANSI Steel SAE 1015 as-rolled'), 'E']

0    207000
Name: E, dtype: int64

# Additional **20 problems** focused on Axis Indexes, Duplicates, `loc`, and `iloc`:

### Axis Index & Index Manipulation

**1.** Set `Material` as the index and sort the DataFrame by the new index.

In [442]:
data = pd.read_csv('Data.csv')

In [443]:
df = pd.DataFrame(data)

In [444]:
df.head()

Std                                ID        Material Heat treatment   Su  \
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   
3  ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD  Steel SAE 1020      as-rolled  448   
4  ANSI  DCE10036FC1946FC8C9108D598D116AD  Steel SAE 1020     normalized  441   

    Sy    A5    Bhn       E      G   mu    Ro     pH Desc  HV  
0  314  39.0  126.0  207000  79000  0.3  7860    NaN  NaN NaN  
1  324  37.0  121.0  207000  79000  0.3  7860    NaN  NaN NaN  
2  284  37.0  111.0  207000  79000  0.3  7860    NaN  NaN NaN  
3  331  36.0  143.0  207000  79000  0.3  7860    NaN  NaN NaN  
4  346  35.8  131.0  207000  79000  0.3  7860  550.0  NaN NaN

In [445]:
df.set_index('Material')

Std                                ID Heat treatment  \
Material                                                                     
Steel SAE 1015       ANSI  D8894772B88F495093C43AF905AB6373      as-rolled   
Steel SAE 1015       ANSI  05982AC66F064F9EBC709E7A4164613A     normalized   
Steel SAE 1015       ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       annealed   
Steel SAE 1020       ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD      as-rolled   
Steel SAE 1020       ANSI  DCE10036FC1946FC8C9108D598D116AD     normalized   
...                   ...                               ...            ...   
Nodular cast iron     JIS  512A80EC21EA416BA2725B38BA8096EF            NaN   
Nodular cast iron     JIS  38526441BA8741CA979DBF870D0B8A9B            NaN   
Nodular cast iron     JIS  CAC03D7EB1AA45E68EFF92A2EF4C3D9B            NaN   
Malleable cast iron   JIS  45C82A36EC644F8BB6170A99ED819B62            NaN   
Malleable cast iron   JIS  BC74F870412F4DDBADDEF1063C1C079F            NaN   

                      Su   Sy    A5    Bhn       E      G    mu    Ro     pH  \
Material                                                                       
Steel SAE 1015       421  314  39.0  126.0  207000  79000  0.30  7860    NaN   
Steel SAE 1015       424  324  37.0  121.0  207000  79000  0.30  7860    NaN   
Steel SAE 1015       386  284  37.0  111.0  207000  79000  0.30  7860    NaN   
Steel SAE 1020       448  331  36.0  143.0  207000  79000  0.30  7860    NaN   
Steel SAE 1020       441  346  35.8  131.0  207000  79000  0.30  7860  550.0   
...                  ...  ...   ...    ...     ...    ...   ...   ...    ...   
Nodular cast iron    600  370   NaN    NaN  169000  70000  0.20  7160  480.0   
Nodular cast iron    700  420   NaN    NaN  169000  70000  0.20  7160  560.0   
Nodular cast iron    800  480   NaN    NaN  169000  70000  0.20  7160  600.0   
Malleable cast iron  400  180   4.0    NaN  160000  64000  0.27  7160  300.0   
Malleable cast iron  500  260   4.0    NaN  160000  64000  0.27  7160  370.0   

                                    Desc     HV  
Material                                         
Steel SAE 1015                       NaN    NaN  
Steel SAE 1015                       NaN    NaN  
Steel SAE 1015                       NaN    NaN  
Steel SAE 1020                       NaN    NaN  
Steel SAE 1020                       NaN    NaN  
...                                  ...    ...  
Nodular cast iron      Nodular cast iron  210.0  
Nodular cast iron      Nodular cast iron  230.0  
Nodular cast iron      Nodular cast iron  240.0  
Malleable cast iron  Malleable cast iron  220.0  
Malleable cast iron  Malleable cast iron  230.0  

[1552 rows x 14 columns]

In [446]:
df.sort_index()

Std                                ID             Material  \
0     ANSI  D8894772B88F495093C43AF905AB6373       Steel SAE 1015   
1     ANSI  05982AC66F064F9EBC709E7A4164613A       Steel SAE 1015   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       Steel SAE 1015   
3     ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD       Steel SAE 1020   
4     ANSI  DCE10036FC1946FC8C9108D598D116AD       Steel SAE 1020   
...    ...                               ...                  ...   
1547   JIS  512A80EC21EA416BA2725B38BA8096EF    Nodular cast iron   
1548   JIS  38526441BA8741CA979DBF870D0B8A9B    Nodular cast iron   
1549   JIS  CAC03D7EB1AA45E68EFF92A2EF4C3D9B    Nodular cast iron   
1550   JIS  45C82A36EC644F8BB6170A99ED819B62  Malleable cast iron   
1551   JIS  BC74F870412F4DDBADDEF1063C1C079F  Malleable cast iron   

     Heat treatment   Su   Sy    A5    Bhn       E      G    mu    Ro     pH  \
0         as-rolled  421  314  39.0  126.0  207000  79000  0.30  7860    NaN   
1        normalized  424  324  37.0  121.0  207000  79000  0.30  7860    NaN   
2          annealed  386  284  37.0  111.0  207000  79000  0.30  7860    NaN   
3         as-rolled  448  331  36.0  143.0  207000  79000  0.30  7860    NaN   
4        normalized  441  346  35.8  131.0  207000  79000  0.30  7860  550.0   
...             ...  ...  ...   ...    ...     ...    ...   ...   ...    ...   
1547            NaN  600  370   NaN    NaN  169000  70000  0.20  7160  480.0   
1548            NaN  700  420   NaN    NaN  169000  70000  0.20  7160  560.0   
1549            NaN  800  480   NaN    NaN  169000  70000  0.20  7160  600.0   
1550            NaN  400  180   4.0    NaN  160000  64000  0.27  7160  300.0   
1551            NaN  500  260   4.0    NaN  160000  64000  0.27  7160  370.0   

                     Desc     HV  
0                     NaN    NaN  
1                     NaN    NaN  
2                     NaN    NaN  
3                     NaN    NaN  
4                     NaN    NaN  
...                   ...    ...  
1547    Nodular cast iron  210.0  
1548    Nodular cast iron  230.0  
1549    Nodular cast iron  240.0  
1550  Malleable cast iron  220.0  
1551  Malleable cast iron  230.0  

[1552 rows x 15 columns]

**2.** Reset the index back to default integer index.

In [447]:
df.reset_index()

index   Std                                ID             Material  \
0         0  ANSI  D8894772B88F495093C43AF905AB6373       Steel SAE 1015   
1         1  ANSI  05982AC66F064F9EBC709E7A4164613A       Steel SAE 1015   
2         2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       Steel SAE 1015   
3         3  ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD       Steel SAE 1020   
4         4  ANSI  DCE10036FC1946FC8C9108D598D116AD       Steel SAE 1020   
...     ...   ...                               ...                  ...   
1547   1547   JIS  512A80EC21EA416BA2725B38BA8096EF    Nodular cast iron   
1548   1548   JIS  38526441BA8741CA979DBF870D0B8A9B    Nodular cast iron   
1549   1549   JIS  CAC03D7EB1AA45E68EFF92A2EF4C3D9B    Nodular cast iron   
1550   1550   JIS  45C82A36EC644F8BB6170A99ED819B62  Malleable cast iron   
1551   1551   JIS  BC74F870412F4DDBADDEF1063C1C079F  Malleable cast iron   

     Heat treatment   Su   Sy    A5    Bhn       E      G    mu    Ro     pH  \
0         as-rolled  421  314  39.0  126.0  207000  79000  0.30  7860    NaN   
1        normalized  424  324  37.0  121.0  207000  79000  0.30  7860    NaN   
2          annealed  386  284  37.0  111.0  207000  79000  0.30  7860    NaN   
3         as-rolled  448  331  36.0  143.0  207000  79000  0.30  7860    NaN   
4        normalized  441  346  35.8  131.0  207000  79000  0.30  7860  550.0   
...             ...  ...  ...   ...    ...     ...    ...   ...   ...    ...   
1547            NaN  600  370   NaN    NaN  169000  70000  0.20  7160  480.0   
1548            NaN  700  420   NaN    NaN  169000  70000  0.20  7160  560.0   
1549            NaN  800  480   NaN    NaN  169000  70000  0.20  7160  600.0   
1550            NaN  400  180   4.0    NaN  160000  64000  0.27  7160  300.0   
1551            NaN  500  260   4.0    NaN  160000  64000  0.27  7160  370.0   

                     Desc     HV  
0                     NaN    NaN  
1                     NaN    NaN  
2                     NaN    NaN  
3                     NaN    NaN  
4                     NaN    NaN  
...                   ...    ...  
1547    Nodular cast iron  210.0  
1548    Nodular cast iron  230.0  
1549    Nodular cast iron  240.0  
1550  Malleable cast iron  220.0  
1551  Malleable cast iron  230.0  

[1552 rows x 16 columns]

**3.** Change the index to the `ID` column and return the first 3 rows using `iloc`.

In [448]:
df.set_index('ID').iloc[:3]

Std        Material Heat treatment   Su  \
ID                                                                           
D8894772B88F495093C43AF905AB6373  ANSI  Steel SAE 1015      as-rolled  421   
05982AC66F064F9EBC709E7A4164613A  ANSI  Steel SAE 1015     normalized  424   
356D6E63FF9A49A3AB23BF66BAC85DC3  ANSI  Steel SAE 1015       annealed  386   

                                   Sy    A5    Bhn       E      G   mu    Ro  \
ID                                                                             
D8894772B88F495093C43AF905AB6373  314  39.0  126.0  207000  79000  0.3  7860   
05982AC66F064F9EBC709E7A4164613A  324  37.0  121.0  207000  79000  0.3  7860   
356D6E63FF9A49A3AB23BF66BAC85DC3  284  37.0  111.0  207000  79000  0.3  7860   

                                  pH Desc  HV  
ID                                             
D8894772B88F495093C43AF905AB6373 NaN  NaN NaN  
05982AC66F064F9EBC709E7A4164613A NaN  NaN NaN  
356D6E63FF9A49A3AB23BF66BAC85DC3 NaN  NaN NaN

**4.** Extract all rows where the current index value is duplicated.

In [449]:
df.index.is_unique

True

In [450]:
df_with_duplicates = pd.concat([df, df.iloc[:3]], ignore_index=False)

In [451]:
df_with_duplicates.head(3)

Std                                ID        Material Heat treatment   Su  \
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   

    Sy    A5    Bhn       E      G   mu    Ro  pH Desc  HV  
0  314  39.0  126.0  207000  79000  0.3  7860 NaN  NaN NaN  
1  324  37.0  121.0  207000  79000  0.3  7860 NaN  NaN NaN  
2  284  37.0  111.0  207000  79000  0.3  7860 NaN  NaN NaN

In [452]:
df_with_duplicates.tail(3)

Std                                ID        Material Heat treatment   Su  \
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   

    Sy    A5    Bhn       E      G   mu    Ro  pH Desc  HV  
0  314  39.0  126.0  207000  79000  0.3  7860 NaN  NaN NaN  
1  324  37.0  121.0  207000  79000  0.3  7860 NaN  NaN NaN  
2  284  37.0  111.0  207000  79000  0.3  7860 NaN  NaN NaN

In [453]:
df_with_duplicates.loc[0]

Std                                ID        Material Heat treatment   Su  \
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   

    Sy    A5    Bhn       E      G   mu    Ro  pH Desc  HV  
0  314  39.0  126.0  207000  79000  0.3  7860 NaN  NaN NaN  
0  314  39.0  126.0  207000  79000  0.3  7860 NaN  NaN NaN

In [454]:
df_with_duplicates.loc[1]

Std                                ID        Material Heat treatment   Su  \
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   

    Sy    A5    Bhn       E      G   mu    Ro  pH Desc  HV  
1  324  37.0  121.0  207000  79000  0.3  7860 NaN  NaN NaN  
1  324  37.0  121.0  207000  79000  0.3  7860 NaN  NaN NaN

In [455]:
df_with_duplicates.loc[2]

Std                                ID        Material Heat treatment   Su  \
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   

    Sy    A5    Bhn       E      G   mu    Ro  pH Desc  HV  
2  284  37.0  111.0  207000  79000  0.3  7860 NaN  NaN NaN  
2  284  37.0  111.0  207000  79000  0.3  7860 NaN  NaN NaN

**5.** Count how many unique values are in the index after setting `Material` as index.

In [456]:
df_with_duplicates.set_index('Material')

Std                                ID Heat treatment  \
Material                                                                     
Steel SAE 1015       ANSI  D8894772B88F495093C43AF905AB6373      as-rolled   
Steel SAE 1015       ANSI  05982AC66F064F9EBC709E7A4164613A     normalized   
Steel SAE 1015       ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       annealed   
Steel SAE 1020       ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD      as-rolled   
Steel SAE 1020       ANSI  DCE10036FC1946FC8C9108D598D116AD     normalized   
...                   ...                               ...            ...   
Malleable cast iron   JIS  45C82A36EC644F8BB6170A99ED819B62            NaN   
Malleable cast iron   JIS  BC74F870412F4DDBADDEF1063C1C079F            NaN   
Steel SAE 1015       ANSI  D8894772B88F495093C43AF905AB6373      as-rolled   
Steel SAE 1015       ANSI  05982AC66F064F9EBC709E7A4164613A     normalized   
Steel SAE 1015       ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       annealed   

                      Su   Sy    A5    Bhn       E      G    mu    Ro     pH  \
Material                                                                       
Steel SAE 1015       421  314  39.0  126.0  207000  79000  0.30  7860    NaN   
Steel SAE 1015       424  324  37.0  121.0  207000  79000  0.30  7860    NaN   
Steel SAE 1015       386  284  37.0  111.0  207000  79000  0.30  7860    NaN   
Steel SAE 1020       448  331  36.0  143.0  207000  79000  0.30  7860    NaN   
Steel SAE 1020       441  346  35.8  131.0  207000  79000  0.30  7860  550.0   
...                  ...  ...   ...    ...     ...    ...   ...   ...    ...   
Malleable cast iron  400  180   4.0    NaN  160000  64000  0.27  7160  300.0   
Malleable cast iron  500  260   4.0    NaN  160000  64000  0.27  7160  370.0   
Steel SAE 1015       421  314  39.0  126.0  207000  79000  0.30  7860    NaN   
Steel SAE 1015       424  324  37.0  121.0  207000  79000  0.30  7860    NaN   
Steel SAE 1015       386  284  37.0  111.0  207000  79000  0.30  7860    NaN   

                                    Desc     HV  
Material                                         
Steel SAE 1015                       NaN    NaN  
Steel SAE 1015                       NaN    NaN  
Steel SAE 1015                       NaN    NaN  
Steel SAE 1020                       NaN    NaN  
Steel SAE 1020                       NaN    NaN  
...                                  ...    ...  
Malleable cast iron  Malleable cast iron  220.0  
Malleable cast iron  Malleable cast iron  230.0  
Steel SAE 1015                       NaN    NaN  
Steel SAE 1015                       NaN    NaN  
Steel SAE 1015                       NaN    NaN  

[1555 rows x 14 columns]

In [457]:
# Count the number of unique index values
unique_material = df_with_duplicates.index.nunique()

In [458]:
unique_material

1552

### Duplicates

**6.** Check for duplicated `Material` values.

In [459]:
df_with_duplicates['is_duplicate_index'] = df_with_duplicates.index.duplicated(keep=False)

In [460]:
df_with_duplicates.groupby(['is_duplicate_index'])['Material'].size()

is_duplicate_index
False    1549
True        6
Name: Material, dtype: int64

In [461]:
df_with_duplicates[df_with_duplicates['is_duplicate_index']]

Std                                ID        Material Heat treatment   Su  \
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   

    Sy    A5    Bhn       E      G   mu    Ro  pH Desc  HV  is_duplicate_index  
0  314  39.0  126.0  207000  79000  0.3  7860 NaN  NaN NaN                True  
1  324  37.0  121.0  207000  79000  0.3  7860 NaN  NaN NaN                True  
2  284  37.0  111.0  207000  79000  0.3  7860 NaN  NaN NaN                True  
0  314  39.0  126.0  207000  79000  0.3  7860 NaN  NaN NaN                True  
1  324  37.0  121.0  207000  79000  0.3  7860 NaN  NaN NaN                True  
2  284  37.0  111.0  207000  79000  0.3  7860 NaN  NaN NaN                True

**7.** Count how many total duplicates are found for the `Material` column.

In [462]:
df_with_duplicates[df_with_duplicates['is_duplicate_index']]['Material'].count()

np.int64(6)

**8.** Keep only the last occurrence of duplicated `Material` values.

In [463]:
df_with_duplicates.drop_duplicates(subset='Material', keep='last')

Std                                ID             Material  \
5     ANSI  2EC038241908434FA714FEEBE24DDEFE       Steel SAE 1020   
8     ANSI  942333E2D11B4C2CA0B9DFD6D1CE38E0       Steel SAE 1022   
12    ANSI  436484F7350147F7A2982D1410FB03CC       Steel SAE 1030   
16    ANSI  C3F57A34049943E98579CCEF761EE90D       Steel SAE 1040   
20    ANSI  60F13C35666C4904BB693C60FFC8297E       Steel SAE 1050   
...    ...                               ...                  ...   
1540   JIS  F7A9289885CD48DC8F7CE65138DD61A2            JIS SUH35   
1545   JIS  48002D4F94F84AC9B4820173CB6B720B       Grey cast iron   
1549   JIS  CAC03D7EB1AA45E68EFF92A2EF4C3D9B    Nodular cast iron   
1551   JIS  BC74F870412F4DDBADDEF1063C1C079F  Malleable cast iron   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       Steel SAE 1015   

         Heat treatment   Su   Sy    A5    Bhn       E      G    mu    Ro  \
5              annealed  395  295  36.5  111.0  207000  79000  0.30  7860   
8              annealed  450  317  35.0  137.0  207000  79000  0.30  7860   
12    tempered at 400 F  848  648  17.0  495.0  207000  79000  0.30  7860   
16    tempered at 400 F  779  593  19.0  262.0  207000  79000  0.30  7860   
20    tempered at 600 F  979  724  14.0  321.0  207000  79000  0.30  7860   
...                 ...  ...  ...   ...    ...     ...    ...   ...   ...   
1540                NaN  880  560   8.0    NaN  206000  80000  0.30  7860   
1545                NaN  350  350   NaN    NaN   85000  34000  0.25  7160   
1549                NaN  800  480   NaN    NaN  169000  70000  0.20  7160   
1551                NaN  500  260   4.0    NaN  160000  64000  0.27  7160   
2              annealed  386  284  37.0  111.0  207000  79000  0.30  7860   

         pH                  Desc     HV  is_duplicate_index  
5       NaN                   NaN    NaN               False  
8       NaN                   NaN    NaN               False  
12      NaN                   NaN    NaN               False  
16      NaN                   NaN    NaN               False  
20      NaN                   NaN    NaN               False  
...     ...                   ...    ...                 ...  
1540    NaN  Heat-resistant steel    NaN               False  
1545  520.0        Grey cast iron  260.0               False  
1549  600.0     Nodular cast iron  240.0               False  
1551  370.0   Malleable cast iron  230.0               False  
2       NaN                   NaN    NaN                True  

[1225 rows x 16 columns]

In [464]:
df_with_duplicates.tail(5)

Std                                ID             Material  \
1550   JIS  45C82A36EC644F8BB6170A99ED819B62  Malleable cast iron   
1551   JIS  BC74F870412F4DDBADDEF1063C1C079F  Malleable cast iron   
0     ANSI  D8894772B88F495093C43AF905AB6373       Steel SAE 1015   
1     ANSI  05982AC66F064F9EBC709E7A4164613A       Steel SAE 1015   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       Steel SAE 1015   

     Heat treatment   Su   Sy    A5    Bhn       E      G    mu    Ro     pH  \
1550            NaN  400  180   4.0    NaN  160000  64000  0.27  7160  300.0   
1551            NaN  500  260   4.0    NaN  160000  64000  0.27  7160  370.0   
0         as-rolled  421  314  39.0  126.0  207000  79000  0.30  7860    NaN   
1        normalized  424  324  37.0  121.0  207000  79000  0.30  7860    NaN   
2          annealed  386  284  37.0  111.0  207000  79000  0.30  7860    NaN   

                     Desc     HV  is_duplicate_index  
1550  Malleable cast iron  220.0               False  
1551  Malleable cast iron  230.0               False  
0                     NaN    NaN                True  
1                     NaN    NaN                True  
2                     NaN    NaN                True

In [465]:
df_with_duplicates.head(5)

Std                                ID        Material Heat treatment   Su  \
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   
3  ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD  Steel SAE 1020      as-rolled  448   
4  ANSI  DCE10036FC1946FC8C9108D598D116AD  Steel SAE 1020     normalized  441   

    Sy    A5    Bhn       E      G   mu    Ro     pH Desc  HV  \
0  314  39.0  126.0  207000  79000  0.3  7860    NaN  NaN NaN   
1  324  37.0  121.0  207000  79000  0.3  7860    NaN  NaN NaN   
2  284  37.0  111.0  207000  79000  0.3  7860    NaN  NaN NaN   
3  331  36.0  143.0  207000  79000  0.3  7860    NaN  NaN NaN   
4  346  35.8  131.0  207000  79000  0.3  7860  550.0  NaN NaN   

   is_duplicate_index  
0                True  
1                True  
2                True  
3               False  
4               False

**9.** Drop duplicates based on `Material` and `Heat treatment`.

In [466]:
df_with_duplicates.drop_duplicates(subset=['Material', 'Heat treatment'])

Std                                ID        Material Heat treatment  \
0     ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled   
1     ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed   
3     ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD  Steel SAE 1020      as-rolled   
4     ANSI  DCE10036FC1946FC8C9108D598D116AD  Steel SAE 1020     normalized   
...    ...                               ...             ...            ...   
1536   JIS  3A8F73A3EC134C5BA72A6597D0D0007D     JIS SUS3093            NaN   
1537   JIS  51BC6B90F8184AFAA0409A8465A5BA9B        JIS SUH1            NaN   
1538   JIS  AC7E5EE7FEAE4945A12C5B7CA7668E43      JIS SUH330            NaN   
1539   JIS  31BA73C9A4E34E4DBF78C950F4C0EA1A      JIS SUH310            NaN   
1540   JIS  F7A9289885CD48DC8F7CE65138DD61A2       JIS SUH35            NaN   

       Su   Sy    A5    Bhn       E      G   mu    Ro     pH  \
0     421  314  39.0  126.0  207000  79000  0.3  7860    NaN   
1     424  324  37.0  121.0  207000  79000  0.3  7860    NaN   
2     386  284  37.0  111.0  207000  79000  0.3  7860    NaN   
3     448  331  36.0  143.0  207000  79000  0.3  7860    NaN   
4     441  346  35.8  131.0  207000  79000  0.3  7860  550.0   
...   ...  ...   ...    ...     ...    ...  ...   ...    ...   
1536  520  205  40.0    NaN  206000  80000  0.3  7860    NaN   
1537  930  685  15.0    NaN  206000  80000  0.3  7860    NaN   
1538  560  205  40.0    NaN  206000  80000  0.3  7860    NaN   
1539  590  205  40.0    NaN  206000  80000  0.3  7860    NaN   
1540  880  560   8.0    NaN  206000  80000  0.3  7860    NaN   

                      Desc  HV  is_duplicate_index  
0                      NaN NaN                True  
1                      NaN NaN                True  
2                      NaN NaN                True  
3                      NaN NaN               False  
4                      NaN NaN               False  
...                    ...  ..                 ...  
1536  Heat-resistant steel NaN               False  
1537  Heat-resistant steel NaN               False  
1538  Heat-resistant steel NaN               False  
1539  Heat-resistant steel NaN               False  
1540  Heat-resistant steel NaN               False  

[1449 rows x 16 columns]

**10.** Find all rows where `Su`, `Sy`, and `E` are all duplicated together.

In [467]:
df_with_duplicates.duplicated(subset=['Su', 'Sy', 'E'])

0       False
1       False
2       False
3       False
4       False
        ...  
1550     True
1551     True
0        True
1        True
2        True
Length: 1555, dtype: bool

In [468]:
# view only the rows where duplication occurs
df_with_duplicates[df_with_duplicates.duplicated(subset=['Su', 'Sy', 'E'])]

Std                                ID             Material  \
115   ANSI  2FD9E4EDBF2244D6BE185B798FB83793      Steel SAE 30202   
117   ANSI  819598B841CD48E3902451D26CCB333E      Steel SAE 30301   
118   ANSI  77A7B865F48C4497B581FFF6BF6F18C4      Steel SAE 30301   
119   ANSI  B4C8C7C886D34158A666A42E042213ED      Steel SAE 30301   
120   ANSI  5E71D9C8D57F4992BE9F12922A94BEE1      Steel SAE 30301   
...    ...                               ...                  ...   
1550   JIS  45C82A36EC644F8BB6170A99ED819B62  Malleable cast iron   
1551   JIS  BC74F870412F4DDBADDEF1063C1C079F  Malleable cast iron   
0     ANSI  D8894772B88F495093C43AF905AB6373       Steel SAE 1015   
1     ANSI  05982AC66F064F9EBC709E7A4164613A       Steel SAE 1015   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       Steel SAE 1015   

     Heat treatment    Su   Sy    A5    Bhn       E       G    mu    Ro  \
115        1/4-hard   862  517  12.0    NaN  193000  743000  0.30  8030   
117        1/4-hard   862  517  25.0    NaN  193000  743000  0.30  8030   
118        1/2-hard  1034  758  15.0    NaN  193000  743000  0.30  8030   
119        3/4-hard  1207  931  12.0    NaN  193000  743000  0.30  8030   
120       Full-hard  1276  965   8.0    NaN  193000  743000  0.30  8030   
...             ...   ...  ...   ...    ...     ...     ...   ...   ...   
1550            NaN   400  180   4.0    NaN  160000   64000  0.27  7160   
1551            NaN   500  260   4.0    NaN  160000   64000  0.27  7160   
0         as-rolled   421  314  39.0  126.0  207000   79000  0.30  7860   
1        normalized   424  324  37.0  121.0  207000   79000  0.30  7860   
2          annealed   386  284  37.0  111.0  207000   79000  0.30  7860   

         pH                 Desc     HV  is_duplicate_index  
115     NaN            stainless    NaN               False  
117     NaN            stainless    NaN               False  
118     NaN            stainless    NaN               False  
119     NaN            stainless    NaN               False  
120     NaN            stainless    NaN               False  
...     ...                  ...    ...                 ...  
1550  300.0  Malleable cast iron  220.0               False  
1551  370.0  Malleable cast iron  230.0               False  
0       NaN                  NaN    NaN                True  
1       NaN                  NaN    NaN                True  
2       NaN                  NaN    NaN                True  

[415 rows x 16 columns]

In [469]:
# show all instances of duplication group
df_with_duplicates[df_with_duplicates.duplicated(subset=['Su', 'Sy', 'E'], keep=False)]

Std                                ID             Material  \
0     ANSI  D8894772B88F495093C43AF905AB6373       Steel SAE 1015   
1     ANSI  05982AC66F064F9EBC709E7A4164613A       Steel SAE 1015   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       Steel SAE 1015   
110   ANSI  C820422DF6594CE48208EE78C63F88C3      Steel SAE 30201   
111   ANSI  082215F3E66545348DCC7F2A8D2D29E0      Steel SAE 30201   
...    ...                               ...                  ...   
1550   JIS  45C82A36EC644F8BB6170A99ED819B62  Malleable cast iron   
1551   JIS  BC74F870412F4DDBADDEF1063C1C079F  Malleable cast iron   
0     ANSI  D8894772B88F495093C43AF905AB6373       Steel SAE 1015   
1     ANSI  05982AC66F064F9EBC709E7A4164613A       Steel SAE 1015   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       Steel SAE 1015   

     Heat treatment    Su   Sy    A5    Bhn       E       G    mu    Ro  \
0         as-rolled   421  314  39.0  126.0  207000   79000  0.30  7860   
1        normalized   424  324  37.0  121.0  207000   79000  0.30  7860   
2          annealed   386  284  37.0  111.0  207000   79000  0.30  7860   
110        1/4-hard   862  517  20.0    NaN  193000  743000  0.30  8030   
111        1/2-hard  1034  758  10.0    NaN  193000  743000  0.30  8030   
...             ...   ...  ...   ...    ...     ...     ...   ...   ...   
1550            NaN   400  180   4.0    NaN  160000   64000  0.27  7160   
1551            NaN   500  260   4.0    NaN  160000   64000  0.27  7160   
0         as-rolled   421  314  39.0  126.0  207000   79000  0.30  7860   
1        normalized   424  324  37.0  121.0  207000   79000  0.30  7860   
2          annealed   386  284  37.0  111.0  207000   79000  0.30  7860   

         pH                 Desc     HV  is_duplicate_index  
0       NaN                  NaN    NaN                True  
1       NaN                  NaN    NaN                True  
2       NaN                  NaN    NaN                True  
110     NaN            stainless    NaN               False  
111     NaN            stainless    NaN               False  
...     ...                  ...    ...                 ...  
1550  300.0  Malleable cast iron  220.0               False  
1551  370.0  Malleable cast iron  230.0               False  
0       NaN                  NaN    NaN                True  
1       NaN                  NaN    NaN                True  
2       NaN                  NaN    NaN                True  

[653 rows x 16 columns]

## A break down with an example and explain what `keep=False` means when using `.duplicated()` in pandas.



### By default:

```python
df.duplicated(subset=['Su', 'Sy', 'E'])
```

Returns `True` **only for the second and subsequent occurrences** of a duplicate combination. The **first occurrence** is considered unique and marked `False`.



### If you use `keep=False`:

```python
df.duplicated(subset=['Su', 'Sy', 'E'], keep=False)
```

This marks **all rows that are part of a duplicate group** — including the first one — as `True`.



### Why it matters:

Let’s say your DataFrame has the following simplified rows:


| index | Su  | Sy  | E      |             |
| ----- | --- | --- | ------ | ----------- |
| 0     | 500 | 300 | 207000 |             |
| 1     | 500 | 300 | 207000 | ← duplicate |
| 2     | 520 | 310 | 207000 |             |
| 3     | 500 | 300 | 207000 | ← duplicate |

#### Using `duplicated(subset=['Su', 'Sy', 'E'])`:

```python
[False, True, False, True]
```

Only rows 1 and 3 are marked as duplicates.

#### Using `duplicated(subset=['Su', 'Sy', 'E'], keep=False)`:

```python
[True, True, False, True]
```

Now **all three rows** with the same combination are marked as `True`.



### Use Case:

If you want to **investigate or drop all rows involved in a duplicate group**, then `keep=False` is the correct option.




### `loc` Practice

**11.** Use `loc` to extract all rows where `Ro` is exactly 7860 and `mu` > 0.25.

In [470]:
df = pd.DataFrame(data)

In [471]:
df.loc[(df['Ro'] == 7860) & (df['mu']) > 0.25]

Std                                ID        Material Heat treatment  \
0     ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled   
1     ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed   
3     ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD  Steel SAE 1020      as-rolled   
4     ANSI  DCE10036FC1946FC8C9108D598D116AD  Steel SAE 1020     normalized   
...    ...                               ...             ...            ...   
1536   JIS  3A8F73A3EC134C5BA72A6597D0D0007D     JIS SUS3093            NaN   
1537   JIS  51BC6B90F8184AFAA0409A8465A5BA9B        JIS SUH1            NaN   
1538   JIS  AC7E5EE7FEAE4945A12C5B7CA7668E43      JIS SUH330            NaN   
1539   JIS  31BA73C9A4E34E4DBF78C950F4C0EA1A      JIS SUH310            NaN   
1540   JIS  F7A9289885CD48DC8F7CE65138DD61A2       JIS SUH35            NaN   

       Su   Sy    A5    Bhn       E      G   mu    Ro     pH  \
0     421  314  39.0  126.0  207000  79000  0.3  7860    NaN   
1     424  324  37.0  121.0  207000  79000  0.3  7860    NaN   
2     386  284  37.0  111.0  207000  79000  0.3  7860    NaN   
3     448  331  36.0  143.0  207000  79000  0.3  7860    NaN   
4     441  346  35.8  131.0  207000  79000  0.3  7860  550.0   
...   ...  ...   ...    ...     ...    ...  ...   ...    ...   
1536  520  205  40.0    NaN  206000  80000  0.3  7860    NaN   
1537  930  685  15.0    NaN  206000  80000  0.3  7860    NaN   
1538  560  205  40.0    NaN  206000  80000  0.3  7860    NaN   
1539  590  205  40.0    NaN  206000  80000  0.3  7860    NaN   
1540  880  560   8.0    NaN  206000  80000  0.3  7860    NaN   

                      Desc  HV  
0                      NaN NaN  
1                      NaN NaN  
2                      NaN NaN  
3                      NaN NaN  
4                      NaN NaN  
...                    ...  ..  
1536  Heat-resistant steel NaN  
1537  Heat-resistant steel NaN  
1538  Heat-resistant steel NaN  
1539  Heat-resistant steel NaN  
1540  Heat-resistant steel NaN  

[692 rows x 15 columns]

**12.** Use `loc` to extract all materials starting with the string “Steel”.

In [472]:
df.loc[df['Material'].str.startswith('Steel')]

Std                                ID                     Material  \
0     ANSI  D8894772B88F495093C43AF905AB6373               Steel SAE 1015   
1     ANSI  05982AC66F064F9EBC709E7A4164613A               Steel SAE 1015   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3               Steel SAE 1015   
3     ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD               Steel SAE 1020   
4     ANSI  DCE10036FC1946FC8C9108D598D116AD               Steel SAE 1020   
...    ...                               ...                          ...   
1360  GOST  75D96A6FC1D14763B0E88ACB8FAC55B2      Steel 40ChL GOST 977-88   
1361  GOST  B4E701BB64DE48E1AB6E32623784454A   Steel 50ChFA GOST 14959-79   
1362  GOST  3D081D531A6F42E08D02BBFC82F0BF8F     Steel 55S2 GOST 14959-79   
1363  GOST  F1F704FCAB7D412B916709CED28F3C1E   Steel 50ChGA GOST 14959-79   
1364  GOST  9C19D505876947D484F8FEBE0348B361  Steel 50ChGFA GOST 14959-79   

                        Heat treatment    Su    Sy    A5    Bhn       E  \
0                            as-rolled   421   314  39.0  126.0  207000   
1                           normalized   424   324  37.0  121.0  207000   
2                             annealed   386   284  37.0  111.0  207000   
3                            as-rolled   448   331  36.0  143.0  207000   
4                           normalized   441   346  35.8  131.0  207000   
...                                ...   ...   ...   ...    ...     ...   
1360  Quenching and cooling in the oil   638   491   NaN    NaN  219000   
1361                      heat treated  1270  1080   8.0    NaN  206000   
1362                      heat treated  1270  1175   6.0    NaN  206000   
1363                      heat treated  1270  1175   7.0    NaN  206000   
1364                      heat treated  1420  1325   6.0    NaN  206000   

          G    mu    Ro     pH               Desc  HV  
0     79000  0.30  7860    NaN                NaN NaN  
1     79000  0.30  7860    NaN                NaN NaN  
2     79000  0.30  7860    NaN                NaN NaN  
3     79000  0.30  7860    NaN                NaN NaN  
4     79000  0.30  7860  550.0                NaN NaN  
...     ...   ...   ...    ...                ...  ..  
1360  85000  0.29  7840    NaN  Carbon cast steel NaN  
1361  80000  0.30  7860    NaN       Spring steel NaN  
1362  80000  0.30  7860    NaN       Spring steel NaN  
1363  80000  0.30  7860    NaN       Spring steel NaN  
1364  80000  0.30  7860    NaN       Spring steel NaN  

[454 rows x 15 columns]

In [473]:
df.loc[df['Material'].str.contains('Steel', case=False)]

Std                                ID                     Material  \
0     ANSI  D8894772B88F495093C43AF905AB6373               Steel SAE 1015   
1     ANSI  05982AC66F064F9EBC709E7A4164613A               Steel SAE 1015   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3               Steel SAE 1015   
3     ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD               Steel SAE 1020   
4     ANSI  DCE10036FC1946FC8C9108D598D116AD               Steel SAE 1020   
...    ...                               ...                          ...   
1360  GOST  75D96A6FC1D14763B0E88ACB8FAC55B2      Steel 40ChL GOST 977-88   
1361  GOST  B4E701BB64DE48E1AB6E32623784454A   Steel 50ChFA GOST 14959-79   
1362  GOST  3D081D531A6F42E08D02BBFC82F0BF8F     Steel 55S2 GOST 14959-79   
1363  GOST  F1F704FCAB7D412B916709CED28F3C1E   Steel 50ChGA GOST 14959-79   
1364  GOST  9C19D505876947D484F8FEBE0348B361  Steel 50ChGFA GOST 14959-79   

                        Heat treatment    Su    Sy    A5    Bhn       E  \
0                            as-rolled   421   314  39.0  126.0  207000   
1                           normalized   424   324  37.0  121.0  207000   
2                             annealed   386   284  37.0  111.0  207000   
3                            as-rolled   448   331  36.0  143.0  207000   
4                           normalized   441   346  35.8  131.0  207000   
...                                ...   ...   ...   ...    ...     ...   
1360  Quenching and cooling in the oil   638   491   NaN    NaN  219000   
1361                      heat treated  1270  1080   8.0    NaN  206000   
1362                      heat treated  1270  1175   6.0    NaN  206000   
1363                      heat treated  1270  1175   7.0    NaN  206000   
1364                      heat treated  1420  1325   6.0    NaN  206000   

          G    mu    Ro     pH               Desc  HV  
0     79000  0.30  7860    NaN                NaN NaN  
1     79000  0.30  7860    NaN                NaN NaN  
2     79000  0.30  7860    NaN                NaN NaN  
3     79000  0.30  7860    NaN                NaN NaN  
4     79000  0.30  7860  550.0                NaN NaN  
...     ...   ...   ...    ...                ...  ..  
1360  85000  0.29  7840    NaN  Carbon cast steel NaN  
1361  80000  0.30  7860    NaN       Spring steel NaN  
1362  80000  0.30  7860    NaN       Spring steel NaN  
1363  80000  0.30  7860    NaN       Spring steel NaN  
1364  80000  0.30  7860    NaN       Spring steel NaN  

[454 rows x 15 columns]

**13.** Use `loc` to change `mu` to 0.31 where `Material` is “Steel SAE 1020 normalized”.

In [474]:
df.loc[
    (df['Material'] == 'Steel SAE 1020') &
    (df['Heat treatment'] == 'normalized'),
    'mu'
] = 0.31

In [475]:
df.loc[
    (df['Material'] == 'Steel SAE 1020') &
    (df['Heat treatment'] == 'normalized')
]

Std                                ID        Material Heat treatment   Su  \
4  ANSI  DCE10036FC1946FC8C9108D598D116AD  Steel SAE 1020     normalized  441   

    Sy    A5    Bhn       E      G    mu    Ro     pH Desc  HV  
4  346  35.8  131.0  207000  79000  0.31  7860  550.0  NaN NaN

**14.** Set all `G` values to 0 where `Sy` < 100 using `loc`.

In [476]:
df = pd.DataFrame(data)

In [477]:
numeric_columns = df.select_dtypes(include=[np.number]).columns
numeric_columns

Index(['Su', 'A5', 'Bhn', 'E', 'G', 'mu', 'Ro', 'pH', 'HV'], dtype='object')

In [478]:
df[numeric_columns] = df[numeric_columns].apply(pd.to_numeric, errors='coerce')

In [481]:
df['Sy'] = pd.to_numeric(df['Sy'], errors='coerce')

In [482]:
df.loc[
    (df['Sy'] < 100),
    'G'
] = 0

In [488]:
df[df['Sy'] < 100][['Sy', 'G']]

Sy  G
175   62.0  0
179   69.0  0
180   69.0  0
183   97.0  0
184   97.0  0
...    ... ..
593   90.0  0
594   55.0  0
1365  50.0  0
1366  75.0  0
1367  90.0  0

[84 rows x 2 columns]

**15.** Filter all rows using `loc` where `Su` is in the top 10% values.

In [489]:
df = pd.DataFrame(data)

In [493]:
df['Su_top_10_percent'] = df['Su'] > df['Su'].quantile(0.90)

In [494]:
df.loc[df['Su_top_10_percent']]

Std                                ID        Material  \
24    ANSI  51BB6C68630743439C326AAF86EA0D7C  Steel SAE 1060   
26    ANSI  E675A2F2E21E40B8830C1B3F23114D89  Steel SAE 1080   
28    ANSI  CE4812A480FD405E8E0DAA0516695E81  Steel SAE 1080   
30    ANSI  5624C3966397415184BEEE0B6D6EEF68  Steel SAE 1095   
32    ANSI  18E85A7791694976A24C8E052BD13769  Steel SAE 1095   
...    ...                               ...             ...   
1513   JIS  EE5CCFBA6A894DC182FF60C8862DA43E        JIS SUP9   
1518   JIS  246BFE5B80D744C8B05EFACE150CF829       JIS SUP10   
1520   JIS  54F6674CB72F465C8D8E1CC5B86D7000     JIS SNCM815   
1521   JIS  A164112854AA482BBC92DAD34B921CA7     JIS SNCM447   
1522   JIS  0B70C2A36C4F43FB81DA06EC5E6B6D74     JIS SNCM630   

         Heat treatment    Su    Sy    A5    Bhn       E      G   mu    Ro  \
24    tempered at 600 F  1103   779  13.0  321.0  207000  79000  0.3  7860   
26           normalized  1010   524  11.0  293.0  207000  79000  0.3  7860   
28    tempered at 600 F  1303   979  12.0  388.0  207000  79000  0.3  7860   
30           normalized  1014   500   9.5  293.0  207000  79000  0.3  7860   
32    tempered at 400 F  1289   827  10.0  401.0  207000  79000  0.3  7860   
...                 ...   ...   ...   ...    ...     ...    ...  ...   ...   
1513       heat treated  1226  1079   9.0    NaN  206000  80000  0.3  7860   
1518                NaN  1226  1079  10.0    NaN  206000  80000  0.3  7860   
1520                NaN  1100   800   9.0    NaN  206000  80000  0.3  7860   
1521                NaN  1100   900  10.0    NaN  206000  80000  0.3  7860   
1522                NaN  1200  1000   9.0    NaN  206000  80000  0.3  7860   

         pH                  Desc  HV  Su_top_10_percent  
24    600.0                   NaN NaN               True  
26      NaN                   NaN NaN               True  
28      NaN                   NaN NaN               True  
30      NaN                   NaN NaN               True  
32      NaN                   NaN NaN               True  
...     ...                   ...  ..                ...  
1513    NaN          Spring steel NaN               True  
1518    NaN          Spring steel NaN               True  
1520    NaN  Case-hardening steel NaN               True  
1521    NaN  Heat-treatment steel NaN               True  
1522    NaN  Heat-treatment steel NaN               True  

[143 rows x 16 columns]

### `iloc` Practice

**16.** Use `iloc` to extract every 5th row between rows 0 and 100.

In [499]:
df = pd.DataFrame(data)
df.head(10)

Std                                ID        Material Heat treatment   Su  \
0  ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015      as-rolled  421   
1  ANSI  05982AC66F064F9EBC709E7A4164613A  Steel SAE 1015     normalized  424   
2  ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3  Steel SAE 1015       annealed  386   
3  ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD  Steel SAE 1020      as-rolled  448   
4  ANSI  DCE10036FC1946FC8C9108D598D116AD  Steel SAE 1020     normalized  441   
5  ANSI  2EC038241908434FA714FEEBE24DDEFE  Steel SAE 1020       annealed  395   
6  ANSI  356B183DD9E34A1C80A5028D43B9E149  Steel SAE 1022      as-rolled  503   
7  ANSI  95CB82FA86314D8490932A9E740744E3  Steel SAE 1022     normalized  483   
8  ANSI  942333E2D11B4C2CA0B9DFD6D1CE38E0  Steel SAE 1022       annealed  450   
9  ANSI  5E035DD0692F47E3A92EB298101AA124  Steel SAE 1030      as-rolled  552   

    Sy    A5    Bhn       E      G    mu    Ro     pH Desc  HV  
0  314  39.0  126.0  207000  79000  0.30  7860    NaN  NaN NaN  
1  324  37.0  121.0  207000  79000  0.30  7860    NaN  NaN NaN  
2  284  37.0  111.0  207000  79000  0.30  7860    NaN  NaN NaN  
3  331  36.0  143.0  207000  79000  0.30  7860    NaN  NaN NaN  
4  346  35.8  131.0  207000  79000  0.31  7860  550.0  NaN NaN  
5  295  36.5  111.0  207000  79000  0.30  7860    NaN  NaN NaN  
6  359  35.0  149.0  207000  79000  0.30  7860    NaN  NaN NaN  
7  359  34.0  143.0  207000  79000  0.30  7860    NaN  NaN NaN  
8  317  35.0  137.0  207000  79000  0.30  7860    NaN  NaN NaN  
9  345  32.0  179.0  207000  79000  0.30  7860    NaN  NaN NaN

In [502]:
df.iloc[0:100:5]

Std                                ID        Material     Heat treatment  \
0   ANSI  D8894772B88F495093C43AF905AB6373  Steel SAE 1015          as-rolled   
5   ANSI  2EC038241908434FA714FEEBE24DDEFE  Steel SAE 1020           annealed   
10  ANSI  8C9BE76E417841C3B821D4776B498039  Steel SAE 1030         normalized   
15  ANSI  CEEFEC39D52C4CBC9FD7A599BF6E4078  Steel SAE 1040           annealed   
20  ANSI  60F13C35666C4904BB693C60FFC8297E  Steel SAE 1050  tempered at 600 F   
25  ANSI  677DEA95360F4ED08E9D8BA76BD75254  Steel SAE 1080          as-rolled   
30  ANSI  5624C3966397415184BEEE0B6D6EEF68  Steel SAE 1095         normalized   
35  ANSI  5412CBE18E014D6CA2349F98FDCB7CF1  Steel SAE 1117           annealed   
40  ANSI  B9823642AF7E4A239F39B049736B33CF  Steel SAE 1137         normalized   
45  ANSI  53A2EC1C4EA24F5AB1E229C4D9E61828  Steel SAE 1141           annealed   
50  ANSI  1EA0C63CE8E646B48BDD0A2CEB4F84AD  Steel SAE 1144  tempered at 400 F   
55  ANSI  849657544E284411BDFD5159978EE22F  Steel SAE 3140           annealed   
60  ANSI  8638842ED9F44FA2AD535A9429DE4BB3  Steel SAE 4130  tempered at 400 F   
65  ANSI  76B0AE8D9FC84B19B8C9777F23F7535C  Steel SAE 4150           annealed   
70  ANSI  51C972EA2EF04A728F9F537C74F69671  Steel SAE 4340           annealed   
75  ANSI  B46E961340E14196B2580DEBF16E0369  Steel SAE 4820           annealed   
80  ANSI  48166F06C0144B878E9104A65AC6F831  Steel SAE 5140  tempered at 400 F   
85  ANSI  23FDAACF3EE5449195F65F16F609D4D6  Steel SAE 5160           annealed   
90  ANSI  2E4BC847E6B54D21BC69AD30393DCFEE  Steel SAE 8620         normalized   
95  ANSI  D18C57467E3D439EA62473DC3D8DB6BA  Steel SAE 8640  tempered at 400 F   

      Su    Sy    A5    Bhn       E      G   mu    Ro     pH Desc  HV  
0    421   314  39.0  126.0  207000  79000  0.3  7860    NaN  NaN NaN  
5    395   295  36.5  111.0  207000  79000  0.3  7860    NaN  NaN NaN  
10   517   345  32.0  149.0  207000  79000  0.3  7860    NaN  NaN NaN  
15   519   353  30.0  149.0  207000  79000  0.3  7860  310.0  NaN NaN  
20   979   724  14.0  321.0  207000  79000  0.3  7860    NaN  NaN NaN  
25   965   586  12.0  293.0  207000  79000  0.3  7860    NaN  NaN NaN  
30  1014   500   9.5  293.0  207000  79000  0.3  7860    NaN  NaN NaN  
35   429   279  32.8  121.0  207000  79000  0.3  7860    NaN  NaN NaN  
40   669   396  22.5  197.0  207000  79000  0.3  7860    NaN  NaN NaN  
45   598   353  25.5  163.0  207000  79000  0.3  7860    NaN  NaN NaN  
50   876   627  17.0  277.0  207000  79000  0.3  7860    NaN  NaN NaN  
55   689   422  24.5  197.0  207000  79000  0.3  7860    NaN  NaN NaN  
60  1627  1462  10.0  467.0  207000  79000  0.3  7860    NaN  NaN NaN  
65   729   379  20.2  197.0  207000  79000  0.3  7860    NaN  NaN NaN  
70   745   472  22.0  217.0  207000  79000  0.3  7860    NaN  NaN NaN  
75   681   464  22.3  197.0  207000  79000  0.3  7860  420.0  NaN NaN  
80  1793  1641   9.0  490.0  207000  79000  0.3  7860  930.0  NaN NaN  
85   722   276  17.2  197.0  207000  79000  0.3  7860    NaN  NaN NaN  
90   633   357  26.3  183.0  207000  79000  0.3  7860    NaN  NaN NaN  
95  1862  1669  10.0  505.0  207000  79000  0.3  7860    NaN  NaN NaN

**17.** Replace the `mu` values of the first 10 rows using `iloc` with linearly increasing values from 0.1 to 0.5.

In [503]:
# generate 10 linearly spaced values between 0.1 and 0.5
new_mu_values = np.linspace(0.1, 0.5, 10)

In [504]:
new_mu_values

array([0.1       , 0.14444444, 0.18888889, 0.23333333, 0.27777778,
       0.32222222, 0.36666667, 0.41111111, 0.45555556, 0.5       ])

In [511]:
df.iloc[:10, df.columns.get_loc('mu')] = new_mu_values.round(2)

In [512]:
df[['Material', 'mu']]

Material    mu
0          Steel SAE 1015  0.10
1          Steel SAE 1015  0.14
2          Steel SAE 1015  0.19
3          Steel SAE 1020  0.23
4          Steel SAE 1020  0.28
...                   ...   ...
1547    Nodular cast iron  0.20
1548    Nodular cast iron  0.20
1549    Nodular cast iron  0.20
1550  Malleable cast iron  0.27
1551  Malleable cast iron  0.27

[1552 rows x 2 columns]

**18.** Use `iloc` to reverse the entire DataFrame order.

In [519]:
df = pd.DataFrame(data)
df.shape

(1552, 15)

In [521]:
df.iloc[::-1]

Std                                ID             Material  \
1551   JIS  BC74F870412F4DDBADDEF1063C1C079F  Malleable cast iron   
1550   JIS  45C82A36EC644F8BB6170A99ED819B62  Malleable cast iron   
1549   JIS  CAC03D7EB1AA45E68EFF92A2EF4C3D9B    Nodular cast iron   
1548   JIS  38526441BA8741CA979DBF870D0B8A9B    Nodular cast iron   
1547   JIS  512A80EC21EA416BA2725B38BA8096EF    Nodular cast iron   
...    ...                               ...                  ...   
4     ANSI  DCE10036FC1946FC8C9108D598D116AD       Steel SAE 1020   
3     ANSI  1C758F8714AC4E0D9BD8D8AE1625AECD       Steel SAE 1020   
2     ANSI  356D6E63FF9A49A3AB23BF66BAC85DC3       Steel SAE 1015   
1     ANSI  05982AC66F064F9EBC709E7A4164613A       Steel SAE 1015   
0     ANSI  D8894772B88F495093C43AF905AB6373       Steel SAE 1015   

     Heat treatment   Su   Sy    A5    Bhn       E      G    mu    Ro     pH  \
1551            NaN  500  260   4.0    NaN  160000  64000  0.27  7160  370.0   
1550            NaN  400  180   4.0    NaN  160000  64000  0.27  7160  300.0   
1549            NaN  800  480   NaN    NaN  169000  70000  0.20  7160  600.0   
1548            NaN  700  420   NaN    NaN  169000  70000  0.20  7160  560.0   
1547            NaN  600  370   NaN    NaN  169000  70000  0.20  7160  480.0   
...             ...  ...  ...   ...    ...     ...    ...   ...   ...    ...   
4        normalized  441  346  35.8  131.0  207000  79000  0.28  7860  550.0   
3         as-rolled  448  331  36.0  143.0  207000  79000  0.23  7860    NaN   
2          annealed  386  284  37.0  111.0  207000  79000  0.19  7860    NaN   
1        normalized  424  324  37.0  121.0  207000  79000  0.14  7860    NaN   
0         as-rolled  421  314  39.0  126.0  207000  79000  0.10  7860    NaN   

                     Desc     HV  
1551  Malleable cast iron  230.0  
1550  Malleable cast iron  220.0  
1549    Nodular cast iron  240.0  
1548    Nodular cast iron  230.0  
1547    Nodular cast iron  210.0  
...                   ...    ...  
4                     NaN    NaN  
3                     NaN    NaN  
2                     NaN    NaN  
1                     NaN    NaN  
0                     NaN    NaN  

[1552 rows x 15 columns]

**19.** Use `iloc` to zero out `Ro` values from rows 50 to 60.

In [522]:
df = pd.DataFrame(data)

In [525]:
df.iloc[50:61, df.columns.get_loc('Ro')] = 0

In [527]:
df[50:61]['Ro']

50    0
51    0
52    0
53    0
54    0
55    0
56    0
57    0
58    0
59    0
60    0
Name: Ro, dtype: int64

**20.** Use `iloc` to swap the values in the first and last rows of the `Sy` column.

In [539]:
df = pd.DataFrame(data)

In [540]:
df['Sy'].head(1)

0    260
Name: Sy, dtype: object

In [541]:
df['Sy'].tail(1)

1551    314
Name: Sy, dtype: object

In [542]:
first_sy = df.iloc[0, df.columns.get_loc('Sy')]

In [543]:
last_sy = df.iloc[-1, df.columns.get_loc('Sy')]

In [544]:
df.iloc[-1, df.columns.get_loc('Sy')] = first_sy

In [545]:
df.iloc[0, df.columns.get_loc('Sy')] = last_sy

In [546]:
df['Sy'].head(1)

0    314
Name: Sy, dtype: object

In [547]:
df['Sy'].tail(1)

1551    260
Name: Sy, dtype: object